File: backend/app/api/v1/news.py

from fastapi import APIRouter, Depends, Request
from sqlalchemy.orm import Session

from app.core.rate_limit import WRITE_LIMIT, limiter
from app.core.security import get_current_user, require_roles
from app.database.session import get_db
from app.enums.user_role import UserRole
from app.models.user_profile import UserProfile
from app.schemas.news import NewsCreate, NewsResponse, NewsUpdate
from app.services import news_service

router = APIRouter(
    prefix="/news",
    tags=["News"]
)

@router.get("/", response_model=list[NewsResponse])
def get_news_feed(
    include_inactive: bool = False,
    limit: int = news_service.DEFAULT_NEWS_LIMIT,
    offset: int = 0,
    db: Session = Depends(get_db),
    current_user: UserProfile = Depends(get_current_user),
):

    if include_inactive and current_user.role not in (UserRole.ADMIN, UserRole.OPERATOR):
        include_inactive = False
    return news_service.list_news(db, include_inactive, limit, offset)

@router.post("/", response_model=NewsResponse, status_code=201)
@limiter.limit(WRITE_LIMIT)
def create_news(
    request: Request,
    payload: NewsCreate,
    db: Session = Depends(get_db),
    current_user: UserProfile = Depends(require_roles(UserRole.ADMIN, UserRole.OPERATOR)),
):
    return news_service.create_news(db, payload, created_by=current_user.id)

@router.get("/{news_id}", response_model=NewsResponse)
def get_news_item(
    news_id: int,
    db: Session = Depends(get_db),
    current_user: UserProfile = Depends(get_current_user),
):
    return news_service.get_news(db, news_id)

@router.patch("/{news_id}", response_model=NewsResponse)
@limiter.limit(WRITE_LIMIT)
def update_news(
    request: Request,
    news_id: int,
    payload: NewsUpdate,
    db: Session = Depends(get_db),
    current_user: UserProfile = Depends(require_roles(UserRole.ADMIN, UserRole.OPERATOR)),
):
    return news_service.update_news(db, news_id, payload)

@router.delete("/{news_id}", status_code=204)
@limiter.limit(WRITE_LIMIT)
def delete_news(
    request: Request,
    news_id: int,
    db: Session = Depends(get_db),
    current_user: UserProfile = Depends(require_roles(UserRole.ADMIN, UserRole.OPERATOR)),
):
    news_service.delete_news(db, news_id)

File: frontend/src/lib/api/news.ts
import api from "@/lib/axios";
import type { NewsCreate, NewsItem, NewsUpdate } from "@/lib/api/types";

export async function getNews(
  includeInactive = false,
  signal?: AbortSignal,
): Promise<NewsItem[]> {
  const { data } = await api.get<NewsItem[]>("/api/v1/news/", {
    params: includeInactive ? { include_inactive: true } : {},
    signal,
  });
  return data;
}

export async function createNews(payload: NewsCreate): Promise<NewsItem> {
  const { data } = await api.post<NewsItem>("/api/v1/news/", payload);
  return data;
}

export async function updateNews(
  newsId: number,
  payload: NewsUpdate,
): Promise<NewsItem> {
  const { data } = await api.patch<NewsItem>(`/api/v1/news/${newsId}`, payload);
  return data;
}

export async function deleteNews(newsId: number): Promise<void> {
  await api.delete(`/api/v1/news/${newsId}`);
}
